### ISSUE TITLE
leverage DCE for compiler/static analysis warnings

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#64219]
Based on #62560 I was wondering if the compiler could leverage the fact that dead code was eliminated to issue compiler or static analysis warnings as it might possibly indicate redundant or even faulty code.

I am not sure if at the stage of DCE there is any information about the actual code in question yet as that is obviously performed on the IR.

```cpp
int multiCompare(const char *haystack)
{
    for (;;) {
        if (*haystack == ' ' || *haystack == '\0') {
            break;
        }
        else {
            do {
                ++haystack;

                if (*haystack == ' ' || *haystack == '\0') {
                    return -1;
                }
                if (*haystack == '|') {
                    break;
                }
            } while (true);

            ++haystack;
        }
    }

    return -1;
}
```

Looking at the code above (which is as of now not treated as dead in Clang) it has a constant result and performs processing of the input pointer without modifying any data at all. Both of those are indicators something might not be right in the code.

CLion reports the constant result via the `Constant Function Result` inspection but that might only indicate that one of the returns *could* be faulty but not the bigger picture that this code does essentially nothing at all.
